c7n/resources/vpc.py

```diff
@@ -2185,6 +2185,64 @@ class resource_type(query.TypeInfo):
         taggable = False
 
 
+@KeyPair.filter_registry.register('unused')
+class UnusedKeyPairs(Filter):
+    """Filters all ec2 keys that are not in use
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: unused-key-pairs
+                resource: aws.key-pair
+                filters:
+                  - unused
+    """
+    annotation_key = 'c7n:unused_keys'
+    permissions = ('ec2:DescribeKeyPairs',)
+    schema = type_schema('unused')
+
+    def process(self, resources, event=None):
+        instances = self.manager.get_resource_manager('ec2').resources()
+        used = set(jmespath.search('[].KeyName', instances))
+        unused = [r for r in resources if r['KeyName'] not in used]
+        return unused
+
+
+@KeyPair.action_registry.register('delete-unused')
+class DeleteUnusedKeyPairs(BaseAction):
+    """Delete all ec2 keys that are not in use
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: delete-unused-key-pairs
+                resource: aws.key-pair
+                filters:
+                  - unused
+                actions:
+                  - delete-unused
+    """
+    permissions = ('ec2:DeleteKeyPair',)
+    schema = type_schema('delete-unused')
+
+    def validate(self):
+        if not [f for f in self.manager.iter_filters() if isinstance(f, UnusedKeyPairs)]:
+            raise PolicyValidationError(
+                "delete-unused should be used in conjunction with unused filter on %s" % (
+                    self.manager.data,))
+        return self
+
+    def process(self, unused):
+        client = local_session(self.manager.session_factory).client('ec2')
+        for key in unused:
+            # not using a try/catch because we want this to raise if the id is invalid
+            client.delete_key_pair(KeyPairId=key['KeyPairId'])
```

Comment:
hmm.. even if it was already deleted? why?

Reply:
you want a separate filter on ec2 for invalid key pair

Reply:
So there is some oddity around the deletes. From what I can see, it looks like as long as the id resembles a key it throws no errors. Even if the key was recently deleted. It will throw an error though if the id is invalid which I would just want to raise. 

<img width="865" alt="Screen Shot 2020-05-07 at 3 51 16 PM" src="https://user-images.githubusercontent.com/13325146/81338579-de682b00-907a-11ea-8d06-635d2f059a3c.png">

<img width="532" alt="Screen Shot 2020-05-07 at 3 51 26 PM" src="https://user-images.githubusercontent.com/13325146/81338560-d7411d00-907a-11ea-83c5-2303ee610475.png">

![image](https://user-images.githubusercontent.com/13325146/81338522-c690a700-907a-11ea-819a-51a2a4fe353f.png)

Reply:
can you associate a `hotdog` key to an ec2 instance? again the resources here are only those that were returned by the describe key pairs api, so under what scenario do you see an invalid key name coming into play?

Reply:
Hmm yeah I guess that is a bogus scenario. I think I was just trying to justify not using the try/catch.. kind of silly. I'll remove the comment.

Reply:
Addressed in 61bbe62